// @aws-sdk/client-amplifyuibuilder/dist-es/models/models_0.js
import { AmplifyUIBuilderServiceException as __BaseException } from "./AmplifyUIBuilderServiceException";
export const CodegenJobGenericDataSourceType = {
    DATA_STORE: "DataStore",
};
export const CodegenGenericDataFieldDataType = {
    AWS_DATE: "AWSDate",
    AWS_DATE_TIME: "AWSDateTime",
    AWS_EMAIL: "AWSEmail",
    AWS_IP_ADDRESS: "AWSIPAddress",
    AWS_JSON: "AWSJSON",
    AWS_PHONE: "AWSPhone",
    AWS_TIME: "AWSTime",
    AWS_TIMESTAMP: "AWSTimestamp",
    AWS_URL: "AWSURL",
    BOOLEAN: "Boolean",
    ENUM: "Enum",
    FLOAT: "Float",
    ID: "ID",
    INT: "Int",
    MODEL: "Model",
    NON_MODEL: "NonModel",
    STRING: "String",
};
export const GenericDataRelationshipType = {
    BELONGS_TO: "BELONGS_TO",
    HAS_MANY: "HAS_MANY",
    HAS_ONE: "HAS_ONE",
};
export const JSModule = {
    ES2020: "es2020",
    ESNEXT: "esnext",
};
export const JSScript = {
    JS: "js",
    JSX: "jsx",
    TSX: "tsx",
};
export const JSTarget = {
    ES2015: "es2015",
    ES2020: "es2020",
};
export const CodegenJobStatus = {
    FAILED: "failed",
    IN_PROGRESS: "in_progress",
    SUCCEEDED: "succeeded",
};
export class InternalServerException extends __BaseException {
    name = "InternalServerException";
    $fault = "server";
    constructor(opts) {
        super({
            name: "InternalServerException",
            $fault: "server",
            ...opts,
        });
        Object.setPrototypeOf(this, InternalServerException.prototype);
    }
}
export class InvalidParameterException extends __BaseException {
    name = "InvalidParameterException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidParameterException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidParameterException.prototype);
    }
}
export class ResourceNotFoundException extends __BaseException {
    name = "ResourceNotFoundException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "ResourceNotFoundException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ResourceNotFoundException.prototype);
    }
}
export class ThrottlingException extends __BaseException {
    name = "ThrottlingException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "ThrottlingException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ThrottlingException.prototype);
    }
}
export const SortDirection = {
    ASC: "ASC",
    DESC: "DESC",
};
export class ResourceConflictException extends __BaseException {
    name = "ResourceConflictException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "ResourceConflictException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ResourceConflictException.prototype);
    }
}
export class ServiceQuotaExceededException extends __BaseException {
    name = "ServiceQuotaExceededException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "ServiceQuotaExceededException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ServiceQuotaExceededException.prototype);
    }
}
export const TokenProviders = {
    FIGMA: "figma",
};
export const FixedPosition = {
    FIRST: "first",
};
export const FormButtonsPosition = {
    BOTTOM: "bottom",
    TOP: "top",
    TOP_AND_BOTTOM: "top_and_bottom",
};
export const FormDataSourceType = {
    CUSTOM: "Custom",
    DATASTORE: "DataStore",
};
export const StorageAccessLevel = {
    PRIVATE: "private",
    PROTECTED: "protected",
    PUBLIC: "public",
};
export const FormActionType = {
    CREATE: "create",
    UPDATE: "update",
};
export const LabelDecorator = {
    NONE: "none",
    OPTIONAL: "optional",
    REQUIRED: "required",
};
export class UnauthorizedException extends __BaseException {
    name = "UnauthorizedException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "UnauthorizedException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, UnauthorizedException.prototype);
    }
}
